drivers/input/input_sbus.c

```diff
@@ -0,0 +1,364 @@
+				     config->channel_info[channel].zephyr_code, value, false,
+				     K_FOREVER);
+			break;
+
+		case INPUT_EV_KEY:
+			if (value > CHANNEL_VALUE_ONE) {
+				input_report_key(dev, config->channel_info[channel].zephyr_code, 1,
+						 false, K_FOREVER);
+			} else if (value < CHANNEL_VALUE_ZERO) {
+				input_report_key(dev, config->channel_info[channel].zephyr_code, 0,
+						 false, K_FOREVER);
+			}
+			break;
+		}
+		data->last_reported_value[channel] = value;
+	}
+}
+
+static void input_sbus_input_report_thread(const struct device *dev, void *dummy2, void *dummy3)
+{
+	struct input_sbus_data *const data = dev->data;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	int i;
+	uint8_t *sbus_channel_data = &data->sbus_frame[1]; /* Omit header */
+	unsigned int channel;
+	unsigned int value;
+	int bitsRead;
+	unsigned int key;
+	int ret;
+	bool connected_reported = false;
+
+	while (true) {
+		if (!data->in_sync) {
+			k_sem_take(&data->report_lock, K_FOREVER);
+			if (data->in_sync) {
+				LOG_DBG("SBUS receiver connected");
+			} else {
+				continue;
+			}
+		} else {
+			ret = k_sem_take(&data->report_lock, K_MSEC(SBUS_INTERFRAME_SPACING));
+			if (ret == -EBUSY) {
+				continue;
+			} else if (ret == -EAGAIN || !data->in_sync) {
+				/* We've lost sync with the UART receiver */
+				key = irq_lock();
+
+				data->partial_sync = false;
+				data->in_sync = false;
+				data->xfer_bytes = 0;
+				irq_unlock(key);
+
+				connected_reported = false;
+				LOG_DBG("SBUS receiver connection lost");
+
+				/* Report connection lost */
+				continue;
+			}
+		}
+
+		if (connected_reported && data->sbus_frame[23] & SBUS_BYTE24_FRAME_LOST) {
+			LOG_DBG("SBUS controller connection lost");
+			connected_reported = false;
+		} else if (!connected_reported &&
+			   !(data->sbus_frame[23] & SBUS_BYTE24_FRAME_LOST)) {
+			LOG_DBG("SBUS controller connected");
+			connected_reported = true;
+		}
+
+		/* Parse the data */
+		channel = 0;
+		value = 0;
+		bitsRead = 0;
+
+		for (i = 0; i < 22; i++) {
+			/* Read the next byte */
+			unsigned char byte = sbus_channel_data[i];
+
+			/* Extract bits and construct the 11-bit value */
+			value |= (byte << bitsRead);
+			bitsRead += 8;
+
+			/* Check if we've read enough bits to form a full 11-bit value */
+			while (bitsRead >= 11) {
+				input_sbus_report(dev, channel, value & 0x7FF);
+
+				/* Shift right to prepare for the next 11 bits */
+				value >>= 11;
+				bitsRead -= 11;
+				channel++;
+			}
+		}
+	}
+}
+
+static void sbus_uart_isr(const struct device *uart_dev, void *user_data)
+{
+	const struct device *dev = user_data;
+	struct input_sbus_data *const data = dev->data;
+
+	if (uart_dev == NULL) {
+		LOG_DBG("UART device is NULL");
+		return;
+	}
+
+	if (!uart_irq_update(uart_dev)) {
+		LOG_DBG("Unable to start processing interrupts");
+		return;
+	}
+
+	while (uart_irq_rx_ready(uart_dev) && data->xfer_bytes <= SBUS_FRAME_LEN) {
+		if (data->in_sync) {
```

Comment:
the block is a bit busy, wondering if some intermediate variable would help making it a bit more comapct, how about one for `data->rd_data`, maybe something else as well

Reply:
unaddressed

Reply:
Decomposed the functionality, `data->rd_data` is compact as it can be unless you want to to track pointer arithmetic in a separate variable but would only introduce potential bugs.

Reply:
I was thinking that you could use a local `uint8_t *rd_data = data->rd_data;`

Reply:
I've added `uint8_t *rd_data = data->rd_data;`